relations/src/shielder/merge.rs

```diff
@@ -0,0 +1,319 @@
+
+        let second_old_trapdoor: FrontendTrapdoor = 23;
+        let second_old_nullifier: FrontendNullifier = 29;
+        let second_old_token_amount: FrontendTokenAmount = 7;
+
+        let new_trapdoor: FrontendTrapdoor = 27;
+        let new_nullifier: FrontendNullifier = 87;
+        let new_token_amount: FrontendTokenAmount = 10;
+
+        let first_old_note = compute_note(
+            token_id,
+            first_old_token_amount,
+            first_old_trapdoor,
+            first_old_nullifier,
+        );
+        let second_old_note = compute_note(
+            token_id,
+            second_old_token_amount,
+            second_old_trapdoor,
+            second_old_nullifier,
+        );
+        let new_note = compute_note(token_id, new_token_amount, new_trapdoor, new_nullifier);
+
+        //                                          merkle root
+        //                placeholder                                        x
+        //        1                       x                     x                       x
+        //   2         3              x        x            x       x              x       x
+        // 4  *5*  ^6^   7          x   x    x   x        x   x   x   x          x   x   x   x
+        //
+        // *first_old_note* | ^second_old_note^
+
+        let zero_note = FrontendNote::default(); // x
+
+        // First Merkle path setup.
+        let first_leaf_index = 5;
+        let first_sibling_note = compute_note(0, 1, 2, 3); // 4
+        let first_parent_note = compute_parent_hash(first_sibling_note, first_old_note); // 2
+
+        // Second Merkle path setup.
+        let second_leaf_index = 6;
+        let second_sibling_note = compute_note(0, 1, 3, 4); // 7
+        let second_parent_note = compute_parent_hash(second_old_note, second_sibling_note); // 3
+
+        // Merkle paths.
+        let first_merkle_path = vec![first_sibling_note, second_parent_note];
+        let second_merkle_path = vec![second_sibling_note, first_parent_note];
+
+        // Common roots.
+        let grandpa_root = compute_parent_hash(first_parent_note, second_parent_note); // 1
+        let placeholder = compute_parent_hash(grandpa_root, zero_note);
+        let merkle_root = compute_parent_hash(placeholder, zero_note);
+
+        MergeRelationWithFullInput::new(
+            MAX_PATH_LEN,
+            token_id,
+            first_old_nullifier,
+            second_old_nullifier,
+            new_note,
+            merkle_root,
+            first_old_trapdoor,
+            second_old_trapdoor,
+            new_trapdoor,
+            new_nullifier,
+            first_merkle_path,
+            second_merkle_path,
+            first_leaf_index,
+            second_leaf_index,
+            first_old_note,
+            second_old_note,
+            first_old_token_amount,
+            second_old_token_amount,
+            new_token_amount,
+        )
+    }
+
+    #[test]
```

Comment:
we don't have negative testcases - maybe we should add some of them to ensure that every of the 4 checks is actually happening